From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: gellert <gellert.hegyi@around.co>
Date: Fri, 13 Aug 2021 10:25:31 +0200
Subject: Add switches to disable Around audio processing, set opus complexity
 and disable audio encoder queue

Add switches to disable Around audio processing, set opus complexity and disable audio encoder queue.

diff --git a/content/browser/renderer_host/render_process_host_impl.cc b/content/browser/renderer_host/render_process_host_impl.cc
index 7caa95301c5e7325b53085029e7c8e2b581acbe3..11cf460679570a971a63c50b6711e1f58fd10485 100644
--- a/content/browser/renderer_host/render_process_host_impl.cc
+++ b/content/browser/renderer_host/render_process_host_impl.cc
@@ -3425,6 +3425,9 @@ void RenderProcessHostImpl::PropagateBrowserCommandLineToRenderer(
     switches::kAndroidFontsPath,
     switches::kAudioBufferSize,
     switches::kAudioLatencyFactor,
+    switches::kDisableAroundAudioProcessing,
+    switches::kOpusComplexity,
+    switches::kDisableEncoderQueue,
     switches::kAutoplayPolicy,
     switches::kMojoCoreLibraryPath,
     switches::kCrossOriginWebAssemblyModuleSharingAllowed,
diff --git a/content/browser/utility_process_host.cc b/content/browser/utility_process_host.cc
index a295f3745dd107e4eb46fc0a58845c4aedbacb52..009e7bdb99444708780a55676fc3d1aa2677f5bc 100644
--- a/content/browser/utility_process_host.cc
+++ b/content/browser/utility_process_host.cc
@@ -269,6 +269,9 @@ bool UtilityProcessHost::StartProcess() {
       // These flags are used by the audio service:
       switches::kAudioBufferSize,
       switches::kAudioLatencyFactor,
+      switches::kDisableAroundAudioProcessing,
+      switches::kOpusComplexity,
+      switches::kDisableEncoderQueue,
       switches::kAudioServiceQuitTimeoutMs,
       switches::kDisableAudioOutput,
       switches::kFailAudioStreamCreation,
diff --git a/media/base/media_switches.cc b/media/base/media_switches.cc
index 271481a22b14f814666957dcd68857856cba3634..d3905726d4411018984ad688a82910b5916290c8 100644
--- a/media/base/media_switches.cc
+++ b/media/base/media_switches.cc
@@ -23,6 +23,15 @@ namespace switches {
 // 0 is default, 1 - 10ms, 2 - 20ms, etc.
 const char kAudioLatencyFactor[] = "audio-latency-factor";
 
+// Disable Around audio processing.
+const char kDisableAroundAudioProcessing[] = "disable-around-audio-processing";
+
+// Sets opus complexity 0-9.
+const char kOpusComplexity[] = "opus-complexity";
+
+// Moves Opus encoding from worker threads to capture thread.
+const char kDisableEncoderQueue[] = "disable-encoder-queue";
+
 // Allow users to specify a custom buffer size for debugging purpose.
 const char kAudioBufferSize[] = "audio-buffer-size";
 
diff --git a/media/base/media_switches.h b/media/base/media_switches.h
index 056cd95a6e7a8365fad52ce8a1869927e6f4f028..2d005552ea971fb9b42b23a3841823d7491ed268 100644
--- a/media/base/media_switches.h
+++ b/media/base/media_switches.h
@@ -23,6 +23,9 @@ class CommandLine;
 namespace switches {
 
 MEDIA_EXPORT extern const char kAudioLatencyFactor[];
+MEDIA_EXPORT extern const char kDisableAroundAudioProcessing[];
+MEDIA_EXPORT extern const char kOpusComplexity[];
+MEDIA_EXPORT extern const char kDisableEncoderQueue[];
 
 MEDIA_EXPORT extern const char kAudioBufferSize[];
 
diff --git a/third_party/blink/renderer/modules/mediastream/media_stream_audio_processor.cc b/third_party/blink/renderer/modules/mediastream/media_stream_audio_processor.cc
index 3b310149b6351190bdfebfefb1de5826bef2491f..db8acd3ba86a8bcf87212d86302ae563e0895cf0 100644
--- a/third_party/blink/renderer/modules/mediastream/media_stream_audio_processor.cc
+++ b/third_party/blink/renderer/modules/mediastream/media_stream_audio_processor.cc
@@ -26,6 +26,7 @@
 #include "media/base/audio_parameters.h"
 #include "media/base/channel_layout.h"
 #include "media/base/limits.h"
+#include "media/base/media_switches.h"
 #include "media/webrtc/helpers.h"
 #include "third_party/abseil-cpp/absl/types/optional.h"
 #include "third_party/blink/public/common/features.h"
@@ -570,6 +571,12 @@ void MediaStreamAudioProcessor::InitializeCaptureFifo(
   // The output channels from the fifo is normally the same as input.
   int fifo_output_channels = input_format.channels();
 
+  if (base::CommandLine::ForCurrentProcess()->HasSwitch(switches::kDisableAroundAudioProcessing)) {
+    auto config = audio_processing_->GetConfig();
+    config.around_audio_processing.enabled = false;
+    audio_processing_->ApplyConfig(config);
+  }
+
   bool aroundAudioProcessingEnabled = audio_processing_->GetConfig().around_audio_processing.enabled;
 
   media::ChannelLayout output_channel_layout;
